[analyzer] MallocChecker – Fix false positive leak for smart pointers in temporary objects

ivanmurashko opened the pull request:
## Summary

Fixes [PR60896](https://github.com/llvm/llvm-project/issues/60896) — false positive leak reports in various smart pointer scenarios including temporaries, inheritance, direct constructor calls, and mixed ownership patterns. Previously, the analyzer had no smart pointer handling in `checkPostCall`, causing it to report false positive leaks for memory properly managed by smart pointers while missing legitimate raw pointer leaks.

## Root Cause and Solution

The original `MallocChecker::checkPostCall()` implementation only handled existing post-call handlers and had no logic for smart pointer scenarios. This caused the analyzer to miss destructor calls for smart pointers in temporary objects, inheritance hierarchies, and direct constructor calls, leading to false positive leak reports for memory that would be properly freed by smart pointer destructors.

The solution implements comprehensive smart pointer detection in `checkPostCall` to identify smart pointer fields in by-value record arguments and direct constructor calls. Add targeted escape logic that only escapes symbols reachable from smart pointer field regions, preserving leak detection for raw pointers. Introduce `EscapeTrackedCallback` class to efficiently mark allocated symbols as escaped when managed by smart pointers. Create helper functions to detect smart pointer types (`unique_ptr`, `shared_ptr`) in both std and custom implementations. Extend field analysis to handle inheritance hierarchies and multiple smart pointer fields. Transform `checkPostCall` from simple post-handler dispatch to unified smart pointer processing while maintaining all existing functionality.

ivanmurashko committed:
[clang-analyzer] Add regression test for PR60896

ivanmurashko committed:
[clang-analizer] MallocChecker: fix false positive leak for unique_ptr in temporary objects

When a unique_ptr is nested inside a temporary object passed by value to a function,
the analyzer couldn't see the destructor call and incorrectly reported a leak.

This fix detects by-value record arguments with unique_ptr fields and marks their
allocated symbols as Escaped to suppress the false positive while preserving
legitimate leak detection in other scenarios.

Key implementation:
- Add isUniquePtrType() to recognize both std::unique_ptr and custom implementations
- Add collectDirectUniquePtrFieldRegions() to scan smart pointer field regions
- Add post-call logic in checkPostCall() to escape allocations from unique_ptr fields
- Handle missing regions with fallback that marks all allocated symbols as escaped

The fix is targeted and safe:
- Only affects by-value record arguments with unique_ptr fields
- Uses proven EscapeTrackedCallback pattern from existing codebase
- Conservative: only suppresses leaks when specific pattern is detected
- Flexible: handles both standard and custom unique_ptr implementations

Fixes PR60896.

ivanmurashko committed:
[clang-analyzer] MallocChecker: extend false positive leak fix to support shared_ptr

Extend the existing post-call escape rule to recognize std::shared_ptr<T> fields
in addition to std::unique_ptr<T>. The fix suppresses false positive leaks
when smart pointers are nested in temporary objects passed by value to functions.

Key changes:
- Replace isUniquePtrType() with isSmartOwningPtrType() that recognizes both
  unique_ptr and shared_ptr (both std:: and custom implementations)
- Update field collection logic to use the generalized smart pointer detection
- Add test coverage for shared_ptr scenarios

The fix remains narrow and safe:
- Only affects rvalue by-value record arguments at call sites
- Only scans from direct smart pointer field regions (no mass escapes)
- Inline-agnostic; no checkPreCall mutations
- Intentionally excludes std::weak_ptr (non-owning)

Fixes PR60896 and extends the solution to cover shared_ptr use cases.

ivanmurashko committed:
[clang-analyzer] Apply clang-format

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+    if (!LooksLikeTemp)
+      continue;
+
+    // Require at least one direct smart owning pointer field by type.
+    const auto *CRD = T->getAsCXXRecordDecl();
+    if (!CRD)
+      continue;
```
How can this fail if we already know its of a record type?
Wouldnt it be simpler to just do this earlier instead of checking the type?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+    const auto *CRD = T->getAsCXXRecordDecl();
+    if (!CRD)
+      continue;
+    bool HasSmartPtrField = false;
+    for (const FieldDecl *FD : CRD->fields()) {
+      if (isSmartOwningPtrType(FD->getType())) {
+        HasSmartPtrField = true;
+        break;
+      }
+    }
+    if (!HasSmartPtrField)
+      continue;
```
This hunk should be hoisted into a dedicated function.
It could use `llvm::any_of` too.
That said, we would again, iterate over the fields later in `collectDirectSmartOwningPtrFieldRegions` if I'm not mistaken. That would again, filter for `isSmartOwningPtrType`.

ivanmurashko replied:
Addressed — the smart pointer field check is now hoisted into a dedicated `hasSmartPtrField` helper using `llvm::any_of`, and we avoid duplicate iteration by only calling `collectDirectSmartOwningPtrFieldRegions` if the check passes.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+      ProgramStateRef State = C.getState();
+      RegionStateTy RS = State->get<RegionState>();
+      ProgramStateRef NewState = State;
+      for (auto [Sym, RefSt] : RS) {
+        if (RefSt.isAllocated() || RefSt.isAllocatedOfSizeZero()) {
+          NewState =
+              NewState->set<RegionState>(Sym, RefState::getEscaped(&RefSt));
+        }
+      }
```
This does something very similar to `EscapeTrackedCallback`. Have you considered reusing some code?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+}
+
 void MallocChecker::checkPostCall(const CallEvent &Call,
                                   CheckerContext &C) const {
+  // Keep existing post-call handlers.
   if (const auto *PostFN = PostFnMap.lookup(Call)) {
     (*PostFN)(this, C.getState(), Call, C);
-    return;
+  }
```
This `checkPostCall` was really simple, and would become a lot more complicated now.
I think we could do better to split things into named functions and separate concerns.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+        ProgramStateRef State = C.getState();
+        RegionStateTy RS = State->get<RegionState>();
+        ProgramStateRef NewState = State;
+        for (auto [Sym, RefSt] : RS) {
+          if (RefSt.isAllocated() || RefSt.isAllocatedOfSizeZero()) {
+            NewState =
+                NewState->set<RegionState>(Sym, RefState::getEscaped(&RefSt));
+          }
+        }
```
There is a lot of code duplication again and again.
Please think about organizing the code to a more maintainable form.

ivanmurashko replied:
The code was refactored and duplication was removed

steakhal commented on clang/test/Analysis/NewDeleteLeaks-PR60896.cpp:
```diff
@@ -0,0 +1,44 @@
+//===----------------------------------------------------------------------===//
+namespace unique_ptr_temporary_PR60896 {
+
+// We use a custom implementation of unique_ptr for testing purposes
+template <typename T>
+struct unique_ptr {
+  T* ptr;
+  unique_ptr(T* p) : ptr(p) {}
+  ~unique_ptr() { delete ptr; }
+  unique_ptr(unique_ptr&& other) : ptr(other.ptr) { other.ptr = nullptr; }
+  T* get() const { return ptr; }
+};
```
I'm pretty sure some header simulator already has a unique_ptr somewhere. We should use that.

ivanmurashko replied:
I looked at system-header-simulator-cxx.h — it has unique_ptr, but without a destructor implementation. For correctness, I kept the custom unique_ptr in the test to ensure proper leak modeling.

steakhal commented on clang/test/Analysis/NewDeleteLeaks-PR60896-shared.cpp:
```diff
@@ -0,0 +1,37 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=core,cplusplus,unix -verify %s
+// expected-no-diagnostics
+
```
Can we combine this test into the unique_ptr one? These two sort of cover the same subjects, so it would make more sense to me to test them in a single file, nearby to each other.

steakhal reviewed (changes requested):
Hey, thanks for the PR. I can see that you spent some time thinking about this issue.
I see no fundamental issues with the proposed changes, except that we could improve the implementation a bit.
Thanks again!

ivanmurashko committed:
[analyzer][test] Refactor smart pointer leak suppression and combine tests

- Applied requested refactoring in MallocChecker (API cleanup, code reuse, duplication reduction, base class field support).
- Merged unique_ptr and shared_ptr PR60896 tests into a single file.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1096,6 +1100,41 @@ class StopTrackingCallback final : public SymbolVisitor {
+///   State->scanReachableSymbols<EscapeTrackedCallback>(RootRegions);
+///   ProgramStateRef NewState = Scan.getState();
+///   if (NewState != State) C.addTransition(NewState);
+class EscapeTrackedCallback final : public SymbolVisitor {
```
Updated as suggested: constructor is now private, added EscapeTrackedRegionsReachableFrom() static helper, and replaced direct visitor usage with the new helper.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+    DC = DC->getParent();
+  }
+  return false;
+}
```
Removed custom isInStdNamespace; now using isWithinStdNamespace from CheckerHelpers.h.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+    // Check if it's in std namespace
+    const DeclContext *DC = ND->getDeclContext();
+    if (!isInStdNamespace(DC))
+      return false;
```
Updated to pass NamedDecl * directly to isWithinStdNamespace

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+        return true;
+      }
+    }
+  }
```
Replaced manual logic with QualType::getAsCXXRecordDecl()

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+        isa<CXXTemporaryObjectExpr>(AE) || isa<MaterializeTemporaryExpr>(AE) ||
+        isa<CXXConstructExpr>(AE) || isa<InitListExpr>(AE) ||
+        isa<ImplicitCastExpr>(AE) ||   // handle common rvalue materializations
+        isa<CXXBindTemporaryExpr>(AE); // handle CXXBindTemporaryExpr
```
Combined into single isa<T1, T2, ...> call

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3107,24 +3111,13 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
-  }
-  return false;
-}
+// Use isWithinStdNamespace from CheckerHelpers.h instead of custom
+// implementation
```
Is this comment still relevant?

ivanmurashko replied:
I removed the comment - not needed anymore

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1096,6 +1098,47 @@ class StopTrackingCallback final : public SymbolVisitor {
+  static ProgramStateRef
+  EscapeTrackedRegionsReachableFrom(ArrayRef<const MemRegion *> Roots,
+                                    ProgramStateRef State) {
+    EscapeTrackedCallback Visitor(State);
+    for (const MemRegion *R : Roots) {
+      State->scanReachableSymbols(loc::MemRegionVal(R), Visitor);
+    }
+    return Visitor.getState();
```
scanReachableSymbols is really expensive. Can't you invoke it only once and just pass Roots to it?

ivanmurashko replied:
Fixed

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1096,6 +1098,47 @@ class StopTrackingCallback final : public SymbolVisitor {
+    return Visitor.getState();
+  }
+
+  ProgramStateRef getState() const { return State; }
```
I suppose we no longer need this.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1096,6 +1098,47 @@ class StopTrackingCallback final : public SymbolVisitor {
+
+  ProgramStateRef getState() const { return State; }
+
+  bool VisitSymbol(SymbolRef Sym) override {
```
This can be in the private part, if we made `SymbolVisitor` friend.

ivanmurashko replied:
I moved `VisitSymbol` to the private part of the class

steakhal reviewed (changes requested):
I like the progress. It's definitely going in the right direction.
Keep it up!

pskrgag reviewed (commented):
Thanks for your work!

I am more C person, so I might be missing some CXX semantics. Except for nits left above patch LGTM.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+  if (!CRD)
+    return;
+
+  for (const FieldDecl *FD : CRD->fields()) {
+    if (!isSmartOwningPtrType(FD->getType()))
+      continue;
+    SVal L = C.getState()->getLValue(FD, loc::MemRegionVal(Base));
+    if (const MemRegion *FR = L.getAsRegion())
+      Out.push_back(FR);
+  }
```
Added base class field support: hasSmartPtrField and collectDirectSmartOwningPtrFieldRegions now recurse into base classes (including virtual). Added base_class_smart_ptr_PR60896 test to verify no false positives for inherited smart pointer fields.

ivanmurashko committed:
[analyzer] MallocChecker: Address minor style and review comments

- Applied minor style fixes and small improvements per review feedback.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3111,174 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+
+  // First try TemplateSpecializationType (for std smart pointers)
+  const auto *TST = QT->getAs<TemplateSpecializationType>();
+  if (TST) {
```
Applied

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -52,6 +52,7 @@
 #include "clang/AST/DeclTemplate.h"
 #include "clang/AST/Expr.h"
 #include "clang/AST/ExprCXX.h"
+
```
Applied

ivanmurashko committed:
[analyzer] MallocChecker: Factor out smart pointer name check

- Moved duplicate "unique_ptr"/"shared_ptr" name check into a local lambda.

ivanmurashko committed:
[clang-analyzer] Fix addTransition misuse - consolidate state updates

Consolidate multiple state transitions into single update to avoid path
splitting. Remove redundant state comparisons and simplify SVal handling.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3111,174 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+    const MemRegion *RCall = VCall.getAsRegion();
+    const MemRegion *RExpr = VExpr.getAsRegion();
+
+    const MemRegion *Base = RCall ? RCall : RExpr;
```
You are absolutely right - that one was fixed

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3111,174 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+      if (NewState != State)
+        C.addTransition(NewState);
+      continue;
+    }
```
I refactored that part of the code and addressed the comment. Especially, I removed the comparison and just set `needsStateUpdate = true` when we know we need a state change

ivanmurashko committed:
[analyzer][test] Add multiple owning arguments test case

Test multiple by-value arguments with smart pointer fields for
comprehensive coverage.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,11 +3110,195 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+      }
+
+      if (hasByValueRecordWithSmartPtr) {
+        State = escapeAllAllocatedSymbols(State);
```
Wow, there is an issue, and the behaviour is not what was expected from the patch. I need to add the corresponding test and fix for the problem.

pskrgag replied:
Thansk!

ivanmurashko committed:
[analyzer] Simplify MallocChecker::checkPostCall

ivanmurashko committed:
[analyzer] scanReachableSymbols is expensive, so we use a single visitor for all roots

pskrgag commented:
I am sorry, looks like I found another regression:

```cpp

#include <memory>

struct Foo {
  std::unique_ptr<int> ptr;
  int *raw;

  Foo() : ptr(std::make_unique<int>(1)), raw(new int) {}
};

void add(Foo foo) {}

int main(int argc, const char **argv) {
  add(Foo());
  return 0;
}
```

With your patch applied `raw` leak is not reported, however current CSA finds it: https://godbolt.org/z/569x11e7a.

I didn't trace why it happens, will be able to come back tomorrow.

ivanmurashko committed:
[analyzer] Fix overly broad escape logic in MallocChecker for mixed ownership scenarios

Fixed the overly broad fallback mechanism in MallocChecker::checkPostCall() that was suppressing legitimate raw pointer leak detection when smart pointers were present. The previous implementation called escapeAllAllocatedSymbols() when it couldn't obtain argument regions for structs with smart pointer fields, which incorrectly escaped all allocated symbols including unrelated raw pointers.

Replaced the broad fallback with targeted escape logic that only handles smart pointer constructor calls, and added precise detection for unique_ptr/shared_ptr constructors that take ownership of pointer arguments. Added comprehensive test coverage for mixed ownership scenarios to ensure raw pointer leaks are detected while smart pointer managed memory is properly handled without false positives.

ivanmurashko committed:
[analyzer] Refactor MallocChecker::checkPostCall to improve readability

Extracted smart pointer processing logic into dedicated helper functions to reduce complexity. The main checkPostCall method is now simplified from 70+ lines to 10 lines while maintaining identical functionality.

ivanmurashko committed:
[analyzer][test] Reorganize NewDeleteLeaks-PR60896 test into unified namespaces

Consolidate test from 7 separate namespaces into 2 unified namespaces to reduce
code duplication and improve maintainability. All test coverage preserved while
reducing file size by 36%.

ivanmurashko committed:
[analyzer] Extract isSmartPtrName helper, add comments, and improve readability

Extract duplicate isSmartPtrName lambda into a static helper function to eliminate
code duplication. Add accurate comments to all helper functions describing their
actual behavior without overstating validation capabilities. Refactor isSmartPtrCall
to use early returns for improved readability.

pskrgag commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,12 +3125,235 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+/// of pointer arguments.
+static bool isSmartPtrCall(const CallEvent &Call) {
+  // Only check for smart pointer constructor calls
+  if (const auto *CD = dyn_cast_or_null<CXXConstructorDecl>(Call.getDecl())) {
```
nit: I think, early returns would make function more readable.

pskrgag reviewed (approved):
Thanks! I've left only one small nit.

LGTM!

ivanmurashko committed:
[analyzer] Fix addTransition API misuse in MallocChecker::checkPostCall

Fix double addTransition calls when existing post-handlers are present.
Previously, checkPostCall would unconditionally call addTransition even
when existing post-handlers (like checkGetDelimOrGetLine) had already
called it, creating multiple transitions from the same exploded graph node.

The fix uses early return when post-handlers exist, ensuring only one
transition per call. Add additional test case with multiple memory owning
arguments (we already had mixed ownership test, this adds multiple separate
arguments) to demonstrate various scenarios that could trigger the issue.

ivanmurashko commented on clang/test/Analysis/NewDeleteLeaks-PR60896.cpp:
```diff
@@ -0,0 +1,80 @@
+  // No warning should be emitted for this - the memory is managed by shared_ptr 
+  // in the temporary Foo object, which will properly clean up the memory
+  add({make_shared<int>(1)});
+}
```
Done! Added test case with multiple memory owning arguments and fixed the `addTransition` API misuse using early return when post-handlers exist. All tests pass.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,12 +3124,240 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
   }
+
+  // Handle smart pointer related processing only if no post-handler was called
+  ProgramStateRef State = handleSmartPointerRelatedCalls(Call, C, C.getState());
+  C.addTransition(State);
```
```suggestion
  C.addTransition(handleSmartPointerRelatedCalls(Call, C, C.getState()));
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,12 +3124,240 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+
+static void collectDirectSmartOwningPtrFieldRegions(
+    const MemRegion *Base, QualType RecQT, CheckerContext &C,
+    SmallVectorImpl<const MemRegion *> &Out) {
```
Shouldnt `out` have set semantics?

ivanmurashko replied:
Good point! I've changed the parameter to use `llvm::SmallPtrSetImpl<const MemRegion *>` to ensure each region is only collected once, preventing duplicates from inheritance traversal (commit a166bb306961).

steakhal reviewed (commented):
I think it looks pretty good. A lot of improvements!

I don't think I have the time to do a proper focused review, so I'd at this point forward you to @NagyDonat to finish the reviews.

steakhal commented:
One more comment. Please don't use force-pushes on GitHub. That ruins the review experience. Use merges instead, and never break your commit history chain.

Another remark would be to add the commit hash to the reply comment where you say that something was fixed. That way the reviewer can easily click on the commit and verify that yes, indeed it was fixed in the right way.

ivanmurashko committed:
[analyzer] Rename function and use set semantics in MallocChecker

- Rename collectDirectSmartOwningPtrFieldRegions to collectSmartOwningPtrFieldRegions
- Use SmallPtrSet instead of SmallVector to prevent duplicates from inheritance traversal

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,12 +3124,240 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+  return false;
+}
+
+static void collectDirectSmartOwningPtrFieldRegions(
```
Good catch! You're absolutely right. I've renamed the function from `collectDirectSmartOwningPtrFieldRegions` to `collectSmartOwningPtrFieldRegions` to better reflect that it now traverses the entire inheritance hierarchy, not just direct fields (commit a166bb306961).

ivanmurashko committed:
[analyzer] Fix out-of-bounds access in handleSmartPointerConstructorArguments

Prevent accessing CD->getParamDecl(I) when I >= CD->getNumParams() for variadic constructors.
Add test case for variadic constructor scenario.

ivanmurashko committed:
Update clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp

fix applied

Co-authored-by: Balazs Benics <benicsbalazs@gmail.com>

ivanmurashko commented:
Thanks for the feedback! I've already added the commit hashes to my responses above. I'll avoid force-pushes going forward and use merges to preserve the commit history.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,12 +3124,242 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+    return isSmartPtrName(ND->getName());
+  }
+
+  // Also try RecordType (for custom smart pointer implementations)
+  if (const auto *RD = QT->getAsCXXRecordDecl()) {
+    // Accept any custom unique_ptr or shared_ptr implementation
+    return isSmartPtrName(RD->getName());
+  }
```
What is the practical purpose of this block?

As far as I see this block returns `true` when it encounters non-template classes named `unique_ptr` or `shared_ptr`, and I don't think that this is an established coding pattern which deserves a separate branch. Is this a block that will never be entered during real-world use of the analyzer or do I misunderstand something here?

(If I understand correctly your tests rely on this branch, but I think it's inelegant to tweak the behavior of the actual software in order to slightly simplify the test code. I would prefer using [or writing] a suitable system header simulator.)

ivanmurashko replied:
Good catch! You're absolutely right. The RecordType fallback became redundant after the constructor-based detection was introduced in bb5eacc79dc69. The constructor mechanism (`handleSmartPointerConstructorArguments`) now properly handles both template and non-template smart pointers, making the RecordType branch unnecessary. That's why I removed it in 08a24effa0007135844fb37d0ecf6e8f1972a45b.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3194,7 +3480,6 @@ void MallocChecker::checkEscapeOnReturn(const ReturnStmt *S,
   if (!Sym)
     // If we are returning a field of the allocated struct or an array element,
     // the callee could still free the memory.
-    // TODO: This logic should be a part of generic symbol escape callback.
```
Is this just a random obsolete TODO, or is it fixed by the this PR? (No action excepted, just curious.)

ivanmurashko replied:
The change is not directly related to the PR. I believe that the TODO comment became obsolete a long time ago, and I deleted it for that reason. 

The TODO comment was added in February 2012 (see 4ca45b1d00e97c590bc1bf012bd35b1f73926565) when fixing a false positive in the malloc checker related to returning fields of allocated structs or array elements via pointer arithmetic. The author of the commit embedded escape-on-return logic directly in `checkPreStmt` but recognized this was architecturally suboptimal, suggesting it should be part of a "generic symbol escape callback" instead.

This TODO was rendered unnecessary when commit 122171e235827 refactored the code by extracting the escape logic into a dedicated `checkEscapeOnReturn` method that could be reused by multiple callbacks (`checkPreStmt` and `checkEndFunction`). This addressed the architectural concern without requiring a complete redesign of the symbol escape callback system.

NagyDonat reviewed (commented):
I reviewed the code changes within this commit, and overall I'm very grateful for this contribution, because it overcomes lots of technical challenges to introduce a promising heuristic.

As this code is unavoidable complex, I think it's especially important to make it _as readable as possible_, so I added several suggestions that try to improve the code quality. I understand that you are probably a bit tired after implementing this impressive commit (and getting through suggestions from other reviewers), so I'm sorry for piling on these additional suggestions. If you feel that some of them are unreasonable, feel free to "push back" -- I'll try to be flexible.

So far I only read the changes in `MallocChecker.cpp`, but I'll also review the tests soon.

Unfortunately I didn't have time to look at the resolved discussion by other reviewers, I'm very sorry if I'm revisiting some topic that was already discussed there.

ivanmurashko committed:
[analyzer] Refactor smart pointer detection and fix naming consistency

Remove redundant RecordType fallback since non-template smart pointers are
already handled by constructor-based detection. Remove std namespace
restrictions for broader coverage and standardize function naming to use
consistent "SmartOwning" prefix for clarity.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,12 +3124,242 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+
+    // Check if it's in std namespace
+    if (!isWithinStdNamespace(ND))
+      return false;
```
I removed the namespace std checks consistently across all smart pointer detection paths to provide broader coverage and reduce false positives, see 08a24effa0007135844fb37d0ecf6e8f1972a45b

NagyDonat reviewed (commented):
Thanks for updating the commit!

I reviewed the tests and added some minor suggestions in the implementation.

Among my earlier suggestions [the visibility of VisitSymbol](https://github.com/llvm/llvm-project/pull/152751/files#r2301055454) and the [complex code duplication question](https://github.com/llvm/llvm-project/pull/152751/files#r2301561854) are still relevant.

Moreover I thought about the approach that you currently emphasize "owning" in every name and comment where you speak about smart pointers. As this is not a distinguishing feature of these functions (you never interact with non-owning smart pointers) and these function names tend to be very long, I think it would be better to omit "owning" from these names. It is enough to mention the exclusion of `weak_ptr` in a single comment (next to the function that recognizes the names of the smart pointer classes).

steakhal commented:
> > > > I don't think I have the time to do a proper focused review, so I'd at this point forward you to @NagyDonat to finish the reviews.
> > > 
> > > 
> > > @steakhal The "Requested change" review status is still active from your old review. Is it just a github glitch?
> > 
> > Im on vacation. I do not track this PR. Merge it if you want.
> 
> Thanks for the information, and sorry for pinging you while on vacation. We'll merge this when the ongoing discussion is finished.

I still cant hold myself from looking at things once or twice a day, but I'm very careful of when I react to something.

ivanmurashko committed:
[analyzer] Eliminate duplicate traversal logic in MallocChecker smart pointer field detection

Refactor hasSmartOwningPtrField and collectSmartOwningPtrFieldRegions to share the same traversal logic using a FieldConsumer helper struct. This eliminates code duplication and ensures consistency between the two functions when traversing record fields and base classes. The change maintains backward compatibility while improving maintainability.

ivanmurashko commented:
Thanks for the review, @NagyDonat. I am still working on addressing the "duplicates the traversal logic" comment and will return to the other comments right after the commit that addresses this major part of the refactoring. Overall, I plan to address all comments by the end of the week and have everything ready for review by Monday.

> 
> I reviewed the tests and added some minor suggestions in the implementation.

Thank you very much. 

> 
> Among my earlier suggestions [the visibility of VisitSymbol](https://github.com/llvm/llvm-project/pull/152751/files#r2301055454) 

Ah, I am very sorry—it looks like my change was missed there. I will definitely add it.

> and the [complex code duplication question](https://github.com/llvm/llvm-project/pull/152751/files#r2301561854) are still relevant.
> 
> Moreover I thought about the approach that you currently emphasize "owning" in every name and comment where you speak about smart pointers. As this is not a distinguishing feature of these functions (you never interact with non-owning smart pointers) and these function names tend to be very long, I think it would be better to omit "owning" from these names. It is enough to mention the exclusion of `weak_ptr` in a single comment (next to the function that recognizes the names of the smart pointer classes).

I also had doubts when choosing between the naming: with or without "Owning." I think it will be better to omit the word "Owning" to keep the names as simple as possible.

ivanmurashko committed:
[analyzer] Make EscapeTrackedCallback::VisitSymbol public

ivanmurashko committed:
Update clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp

Updated comment for isSmartOwningPtrName

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

ivanmurashko committed:
Update clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

ivanmurashko committed:
Update clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

ivanmurashko committed:
Update clang/test/Analysis/NewDeleteLeaks-PR60896.cpp

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

ivanmurashko committed:
Update clang/test/Analysis/NewDeleteLeaks-PR60896.cpp

LIT tests updates to reflect the heuristic behaviour in the tests

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

ivanmurashko committed:
LIT tests updates to reflect the heuristic behaviour in the tests (shared_ptr update)

ivanmurashko committed:
[analyzer] Refactoring (name change): SmartOwningPtr -> SmartPtr

ivanmurashko committed:
[analyzer] Rename Base variables in MallocChecker to avoid confusion with getBaseRegion()

Rename parameter and member variable 'Base' to 'Reg' and local variable 'BaseRegion' to 'BaseObjRegion' in collectSmartPtrFieldRegions and FieldConsumer. This clarifies that these variables refer to general memory regions and base class object regions respectively, not the MemRegion::getBaseRegion() method which has different semantics.

ivanmurashko committed:
[analyzer] Move variadic constructor test to separate namespace and use unique_ptr

Move Test 8 to a dedicated variadic_constructor_tests namespace and rename VariadicSmartPtr to unique_ptr to ensure it's recognized by the smart pointer name checking logic. This change isolates the variadic constructor bounds testing scenario and activates the specific MallocChecker logic for testing out-of-bounds access when Call.getNumArgs() > CD->getNumParams().

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3068,12 +3124,225 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+}
+
+static void collectSmartOwningPtrFieldRegions(
+    const MemRegion *Base, QualType RecQT, CheckerContext &C,
```
Good point. I applied the name change in 49b7d31d174559b343d83a3629b35e03a36d96e0.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1096,6 +1104,54 @@ class StopTrackingCallback final : public SymbolVisitor {
+
+  explicit EscapeTrackedCallback(ProgramStateRef S) : State(std::move(S)) {}
+
+  bool VisitSymbol(SymbolRef Sym) override {
```
That's my bad. I hope that I finally addressed that in 64d3aef5bf2058ed6ba92cbdc6ce9ce362de8900

ivanmurashko commented:
> Moreover I thought about the approach that you currently emphasize "owning" in every name and comment where you speak about smart pointers. As this is not a distinguishing feature of these functions (you never interact with non-owning smart pointers) and these function names tend to be very long, I think it would be better to omit "owning" from these names. It is enough to mention the exclusion of `weak_ptr` in a single comment (next to the function that recognizes the names of the smart pointer classes).

I eliminated the "owning" word in commit 692db9e74c0cd11e24d01777d28849151250edd7

NagyDonat commented on clang/test/Analysis/NewDeleteLeaks-PR60896.cpp:
```diff
@@ -187,12 +202,12 @@ void test_variadic_constructor_bounds() {
+  process_variadic_smart_ptr(unique_ptr<int>(raw_ptr, 1, 2, 3, 4, 5));
   
   (void)malloc_ptr;
 } // expected-warning {{Potential leak of memory pointed to by 'malloc_ptr'}} expected-note {{Potential leak of memory pointed to by 'malloc_ptr'}}
```
```suggestion
} // expected-warning {{Potential leak of memory pointed to by 'malloc_ptr'}}
  // expected-note@-1 {{Potential leak of memory pointed to by 'malloc_ptr'}}
```
It is possible to move an `expected-` markers onto a different line (by adding `@` + relative line number) and it may be a nice touch on long lines like this.

NagyDonat reviewed (approved):
Thanks for all the updates!

I think the change is basically ready to merge, but I added two very minor suggestions in inline comments.

ivanmurashko committed:
Update clang/test/Analysis/NewDeleteLeaks-PR60896.cpp

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

ivanmurashko committed:
[analyzer] Document dual behavior of hasSmartPtrField function

Add comment explaining that when FieldConsumer is provided, the function
always returns false but performs collection as a side effect.

ivanmurashko commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3145,27 +3144,71 @@ static bool isSmartOwningPtrType(QualType QT) {
 /// Check if a record type has smart owning pointer fields (directly or in base
-/// classes).
-static bool hasSmartOwningPtrField(const CXXRecordDecl *CRD) {
+/// classes). When FC is provided, also collect the field regions.
```
@ NagyDonat, Good point! Added documentation explaining this dual behavior in 196abdb3e4bd4eced37933ddf2d351bf3a39320e. When FC is provided, the return value should be ignored since it always returns false for full traversal.

NagyDonat replied:
Thanks!
